docs/source/notes/numpy_compatibility.rst

```diff
@@ -0,0 +1,221 @@
+    "``int$``", "``int$``", ":math:`8, 16, 32, 64`"
+    "``float$``", "``float$``", ":math:`16, 32, 64`"
+    "``complex$``", "``complex$``", ":math:`64, 128`"
+
+.. warning::
+
+   The promotion and casting rules `of NumPy`_ differ from those `of PyTorch`_.
+
+To ``numpy``
+^^^^^^^^^^^^^
+
+The :meth:`torch.numpy()` method  and the :doc:`np.asarray()
+<numpy:reference/generated/numpy.asarray>` function returns a **view** of the
+underlying tensor as a ``np.ndarray`` object.
+
+.. code-block:: python
+
+    b_torch = torch.ones(3)
+    b_torch.numpy()[2] = 32
+    # b_torch is tensor([ 1.,  1., 32.])
+    a_np = np.array([1, 1, 32], dtype = np.float32)
+    np.array_equal(b_torch.numpy() == a_np) # True
+    c_tmp = np.asarray(b_torch, dtype = np.float32) # No copy if same dtype
+    # c_tmp is array([1., 1., 32.], dtype=float32)
+    c_tmp[2] = 1.
+    # b_torch is tensor([1., 1., 1.])
+
+.. note::
+
+   Since ``np.asarray()`` depends on the implementation of
+   ``torch.Tensor.__array__()`` which calls ``torch.numpy()``, the **leaf node**
+   requirement still needs to be satisfied by the user, i.e., ``requires_grad ==
+   False``
+
+From ``numpy``
+^^^^^^^^^^^^^^^^^
+
+To obtain a **view** of the data, :meth:`torch.from_numpy()` can be used.
+
+.. code-block:: python
+
+   import torch
+   import numpy as np
+   a_np = np.array([1, 2, 3], dtype = np.float64)
+   b_torch = torch.from_numpy(a_np)
+   # b_torch = torch.as_tensor(a_np) # see note
+   b_torch[2] = 23
+   assert(b_torch.numpy() == a_np) # True
+
+- :meth:`torch.from_numpy()` is guaranteed to share memory with NumPy.
+- :meth:`torch.as_tensor()` will try to stay away from copy operations, it
+  also has the effect of sharing memory. However, ``torch.as_tensor()`` has
+  slightly higher overhead as it checks and accepts other iteratable objects as
+  well, e.g. ``list`` objects.
+- :meth:`torch.from_dlpack()` called with a NumPy array as its argument will also generate a ``torch.Tensor`` view.
+
+To obtain a **copy** of the ``ndarray`` object, and not share memory, the
+:meth:`torch.tensor()` constructor accepts :meth:`np.ndarray` objects as a data
+source to construct and return a ``torch.Tensor``.
+
+.. note::
+
+   Recall that, if ``x`` is a tensor, ``torch.tensor(x)`` is equivalent to
+    ``x.clone().detach()``.
+
+.. code-block:: python
+
+   import torch
+   import numpy as np
+   a_np = np.array([1, 2, 3], dtype = np.float64)
+   b_torch = torch.tensor(a_np)
+   # b_torch.dtype is torch.float64
+   b_torch[2] = 23
+   assert(b_torch.numpy() == a_np) # Will fail
+
+Special considerations apply when calling NumPy functions on PyTorch tensors.
+Where possible, the equivalent PyTorch function is called.
+
+.. note::
+
+   This is due to the fact that ``torch.Tensor.__array_priority__`` is higher
+   than the NumPy default of ``0``.
+
+Indexing
+^^^^^^^^^^^^
+
+For the most part, both simple and "fancy" indexing work as expected. Edge cases
+are enumerated here.
```

Comment:
The main point here was that one can index a tensor with an ndarray, and vice versa. Can you state that, with a few examples? And again a recommendation that explicit is better than implicit - work in one library as a user and do the conversions up front.

Reply:
Yup, gave it a shot.